### ISSUE TITLE
Unexpected behavior when reflecting on constructor that has parameter that sets instance variable using syntactic sugar

### ISSUE THREAD

**Issue by fertrig** [dart-lang/reflectable#170]
Given the constructor below which uses syntactic sugar to assign a constructor argument to an instance variable.
```
class Sugar {
  Sugar({this.foo = 222});

  final int foo;
}
```
When I invoke the constructor using `newInstance` without any arguments, then the default value of the constructor parameter does not seem to be used. In the example above, after calling `newInstance` without any arguments, the instance variable `foo` is equal to `null`. I'd expect the instance variable `foo` to be equal to `222` after calling `newInstance` without any arguments.

Also, when I get the default value of the constructor parameter, it returns `null`. I'd expect it to return `222`.

I believe the issue is around the syntactic sugar used in the constructor parameter.

I wrote a test to show the failures based on my expectations:
https://gist.github.com/fertrig/8b1c50deb09c6422779fe5014bdd8aeb

Are my expectations correct? Is there a workaround?

**Comment by eernstg** [comment]
Sounds like a bug. (I'm on vacation, but I'll take a look next week).

**Comment by eernstg** [comment]
So quickly a week can evaporate when part of it is vacation. ;-)

The case where a named constructor argument has a default value which is used (because no parameter with that name is passed at a call site) wasn't covered by the current tests. But I just added it ([here](https://github.com/dart-lang/test_reflectable/blob/f635327aa3f56d34c4c8381bcbe7fd68ef4a0705/test/new_instance_default_values_test.dart#L50)), and that test does succeed.

**Comment by eernstg** [comment]
Turns out that an initializing formal constructor parameter (`this.something`) which is optional is _not_ of type `DefaultFormalParameter` in the Dart analyzer program model. So these parameters are not considered when the default value code is generated for the other optional parameters.

I don't know how to obtain that default value at this point (or even how to discover that a given initializing formal parameter has a default value), so I'll have to return to this later.

**Comment by eernstg** [comment]
Awaiting a review, you may be able to use the fix in the branch 'fix_170_may19' by putting this into pubspec.yaml:
```
  reflectable:
    git:
      url: git://github.com/dart-lang/reflectable.git
      ref: fix_170_may19
```
This is expected to be published as 2.0.11.

**Comment by eernstg** [comment]
Closing: The branch is now gone and 2.0.11 has been published.

**Comment by fertrig** [comment]
Thank you so much, @eernstg!

Invoking the constructor now works as expected. 

However, when I inspect the parameter with the initializing formal, its parameter mirror reports that it has a default value but it also says that its default value is null.

If a constructor parameter is initialized like this?
```
Foo({ this.bar = 222 });
```
Is the parameter `bar` considered to have a default value? Its parameter mirror is currently reporting that `hasDefaultValue` is true, but that `defaultValue` is null. This seems inconsistent.

One could also argue that `hasDefaultValue` should be false since this initializer is syntactic sugar.

Here's is a gist with a test that fails showing the issue:
https://gist.github.com/fertrig/8b1c50deb09c6422779fe5014bdd8aeb

**Comment by eernstg** [comment]
Thanks, @fertrig!

Argh, that is another bug in the same area, `defaultValue` should return 222. A fix is available in branch `fix_170b_may19`, to be published as 2.0.12.

**Comment by eernstg** [comment]
2.0.12 was just published (and branch `fix_170b_may19` was deleted).
